procedures/unit-testing-junit.ipf

```diff
@@ -166,119 +73,141 @@ static Function/S JU_TrimSOUT(input, [listSepStr])
-	if(strlen(juTC.systemOut))
-		sout += "\t\t<system-out>" + JU_ToXMLCharacters(JU_TrimSOUT(juTC.systemOut)) + "</system-out>\n"
+static Function/S JU_CaseToOut(testSuiteIndex, testCaseIndex)
+	variable testSuiteIndex, testCaseIndex
+
+	string out, name, classname, message, type
+	variable i, timeTaken, startIndex, endIndex
+
+	WAVE/T wvTestSuite = UTF_Reporting#GetTestSuiteWave()
+	WAVE/T wvTestCase = UTF_Reporting#GetTestCaseWave()
+	WAVE/T wvAssertion = UTF_Reporting#GetTestAssertionWave()
+
+	classname = JU_ToXMLToken(JU_ToXMLCharacters(wvTestCase[testCaseIndex][%NAME]))
+	name = JU_ToXMLToken(JU_ToXMLCharacters(wvTestSuite[testSuiteIndex][%PROCEDURENAME]))
+	sprintf name, "%s in %s (%d)", classname, name, testCaseIndex
+	timeTaken = str2num(wvTestCase[testCaseIndex][%ENDTIME]) - str2num(wvTestCase[testCaseIndex][%STARTTIME])
+
+	sprintf out, "\t\t<testcase name=\"%s\" classname=\"%s\" time=\"%.3f\">\n", name, classname, timeTaken
+	if(!CmpStr(IUTF_STATUS_SKIP, wvTestCase[testCaseIndex][%STATUS]))
+		out += "\t\t\t<skipped/>\n"
 	endif
-	if(strlen(juTC.systemErr))
-		sout += "\t\t<system-err>" + JU_ToXMLCharacters(juTC.systemErr) + "</system-err>\n"
+
+	startIndex = str2num(wvTestCase[testCaseIndex][%CHILD_START])
+	endIndex = str2num(wvTestCase[testCaseIndex][%CHILD_END])
+	for(i = startIndex; i < endIndex; i += 1)
+		message = JU_ToXMLCharacters(wvAssertion[i][%MESSAGE])
+		type = JU_ToXMLCharacters(wvAssertion[i][%TYPE])
+		// we are outputing everything as error to keep the same behavior as older versions of UTF
+
+		// strswitch(wvAssertion[i][%TYPE])
+		// 	case IUTF_STATUS_FAIL:
+		// 		s += "\t\t\t<failure message=\"" + message + "\" type=\"" + type + "\"></failure>\n"
+		// 		break
+		// 	case IUTF_STATUS_ERROR:
+				out += "\t\t\t<error message=\"" + message + "\" type=\"" + type + "\"></error>\n"
+		// 		break
+		// 	default:
+		// 		break
+		// endswitch
+	endfor
+
+	message = wvTestCase[testCaseIndex][%STDOUT]
+	if(strlen(message))
+		out += "\t\t<system-out>" + JU_ToXMLCharacters(JU_TrimSOUT(message)) + "</system-out>\n"
+	endif
+	message = wvTestCase[testCaseIndex][%STDERR]
+	if(strlen(message))
+		out += "\t\t<system-err>" + JU_ToXMLCharacters(message) + "</system-err>\n"
 	endif
 
-	return (sout + "\t\t</testcase>\n")
+	return (out + "\t\t</testcase>\n")
 End
 
-/// Adds a JUNIT Test Suite property to the list of properties for current Suite
-static Function JU_AddTSProp(juTSProp, propName, propValue)
-	STRUCT strSuiteProperties &juTSProp
-	string propName
-	string propValue
-
-	if(!UTF_Utils#IsEmpty(propName))
-		propName = JU_ToXMLToken( JU_ToXMLCharacters(propName))
-		propValue = JU_ToXMLCharacters(propValue)
-		juTSProp.propNameList = AddListItem(propName, juTSProp.propNameList, "<")
-		juTSProp.propValueList = AddListItem(propValue, juTSProp.propValueList, "<")
-	endif
+/// Converts the reference time string that is stored in the result storage waves into an absolute
+/// time which counts the seconds since 1904-01-01.
```

Comment:
Shouldn't it be reference + now - num ?
I don't like that you assume here that DateTime and StopMSTimer use the same absolute time source, but there is probably no better way in Igor.

Reply:
From my understanding `StopMSTimer(-2) * IUTF_MICRO_TO_ONE` counts the seconds since computer startup which is some kind fixed point in time. DateTime counts the seconds since another fixed point in time. Calculating `DateTime - StopMSTimer(-2) * IUTF_MICRO_TO_ONE` returns you the difference between these fixed points in time.

`num` is a time in the `StopMSTimer(-2)` timeline and needs to be converted into the `DateTime` timeline. To do this I have to add the previous difference between both fixed points in time to `num` which results in:

```igorpro
variable result = num + DateTime - StopMSTimer(-2) * IUTF_MICRO_TO_ONE
```

So I think my formula is still correct.

Reply:
@MichaelHuth 

> I don't like that you assume here that DateTime and StopMSTimer use the same absolute time source, but there is probably no better way in Igor.

We could track duration and start time separately so that we don't have to mix the two different time sources. Do you think this is a good idea?

Reply:
For this I have to add a new column to track the start time in a different format. I cannot use `%STARTTIME` for this as this is used to calculate the time difference later when comparing to `%ENDTIME`.

I can store the value in the new column as ISO8601 as most consumers that will take use of this value will convert it to ISO8601 anyway. So I think we could name the column `%STARTTIME_ISO8601`.

Reply:
Adding an `%ENDTIME_ISO8601` has no use as we do not have any consumers that will take use of it.

Reply:
Ok, so what bugs me here is that the time difference calculation and the conversion to absolute time is mixed in the same formula. I think that should be separate and the conversion to absolute time a simple 
`time_point + time_difference`.

I think in the end we output only ms precision. 
The actual precision of DateTime can be less than ms.

Reply:
I have updated the code documentation to make it more understandable what is being done.

Reply:
@MichaelHuth If this is good from your POV now, please approve. Otherwise please suggest a change.

Reply:
LGTM